File: code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/AssuranceConstants.java
/*
 * Copyright 2022 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

package com.adobe.marketing.mobile.assurance;


import java.util.HashMap;
import java.util.Map;

final class AssuranceConstants {
    static final String VENDOR_ASSURANCE_MOBILE = "com.adobe.griffon.mobile";

    static final class DeeplinkURLKeys {
        static final String START_URL_QUERY_KEY_SESSION_ID = "adb_validation_sessionid";
        static final String START_URL_QUERY_KEY_ENVIRONMENT = "env";

        private DeeplinkURLKeys() {}
    }

    static final class GenericEventPayloadKey {
        static final String ACP_EXTENSION_EVENT_TYPE = "ACPExtensionEventType";
        static final String ACP_EXTENSION_EVENT_SOURCE = "ACPExtensionEventSource";
        static final String ACP_EXTENSION_EVENT_NAME = "ACPExtensionEventName";
        static final String ACP_EXTENSION_EVENT_DATA = "ACPExtensionEventData";
        static final String ACP_EXTENSION_EVENT_UNIQUE_IDENTIFIER =
                "ACPExtensionEventUniqueIdentifier";
        static final String ACP_EXTENSION_EVENT_NUMBER = "ACPExtensionEventNumber";

        private GenericEventPayloadKey() {}
    }

    static final class SDKEventDataKey {
        static final String START_SESSION_URL = "startSessionURL";
        static final String EXTENSIONS = "extensions";
        static final String STATE_OWNER = "stateowner";
        static final String FRIENDLY_NAME = "friendlyName";

        private SDKEventDataKey() {}
    }

    static final class SDKEventName {
        static final String XDM_SHARED_STATE_CHANGE = "Shared state change (XDM)";
        static final String SHARED_STATE_CHANGE = "Shared state change";

        private SDKEventName() {}
    }

    static final class SDKSharedStateName {
        static final String CONFIGURATION = "com.adobe.module.configuration";
        static final String EVENTHUB = "com.adobe.module.eventhub";

        private SDKSharedStateName() {}
    }

    static final class SDKConfigurationKey {
        static final String ORG_ID = "experienceCloud.org";

        private SDKConfigurationKey() {}
    }

    static final class AssuranceEventType {
        static final String GENERIC = "generic";
        static final String LOG = "log";
        static final String CONTROL = "control";
        static final String CLIENT = "client";
        static final String BLOB = "blob";

        private AssuranceEventType() {}
    }

    static final class ControlType {
        static final String START_EVENT_FORWARDING = "startEventForwarding";
        static final String SCREENSHOT = "screenshot";
        static final String LOG_FORWARDING = "logForwarding";
        static final String FAKE_EVENT = "fakeEvent";
        static final String CONFIG_UPDATE = "configUpdate";
        static final String NONE = "none";
        static final String WILDCARD = "wildcard";

        private ControlType() {}
    }

    static final class AssuranceEventKeys {
        static final String EVENT_ID = "eventID";
        static final String VENDOR = "vendor";
        static final String TYPE = "type";
        static final String PAYLOAD = "payload";
        static final String METADATA = "metadata";
        static final String TIMESTAMP = "timestamp";
        static final String EVENT_NUMBER = "eventNumber";
        static final String CHUNK_DATA = "chunkData";
        static final String CHUNK_ID = "chunkId";
        static final String CHUNK_TOTAL = "chunkTotal";
        static final String CHUNK_SEQUENCE_NUMBER = "chunkSequenceNumber";

        private AssuranceEventKeys() {}
    }

    static final class SharedStateKeys {
        static final String ASSURANCE_STATE_SESSION_ID = "sessionid";
        static final String ASSURANCE_STATE_CLIENT_ID = "clientid";
        static final String ASSURANCE_STATE_INTEGRATION_ID = "integrationid";

        private SharedStateKeys() {}
    }

    static final class DataStoreKeys {
        static final String DATASTORE_NAME = "com.adobe.assurance.preferences";
        static final String SESSION_URL = "reconnection.url";
        static final String ENVIRONMENT = "environment";
        static final String CLIENT_ID = "clientid";
        static final String SESSION_ID = "sessionid";
        static final String TOKEN = "token";

        private DataStoreKeys() {}
    }

    static final class BlobKeys {
        static final String UPLOAD_ENDPOINT_FORMAT = "https://blob%s.griffon.adobe.com";
        static final String UPLOAD_PATH_API = "api";
        static final String UPLOAD_PATH_FILEUPLOAD = "FileUpload";
        static final String UPLOAD_QUERY_KEY = "validationSessionId";
        static final String UPLOAD_HTTP_METHOD = "POST";
        static final String UPLOAD_HEADER_KEY_CONTENT_TYPE = "Content-Type";
        static final String UPLOAD_HEADER_KEY_FILE_CONTENT_TYPE = "File-Content-Type";
        static final String UPLOAD_HEADER_KEY_CONTENT_LENGTH = "Content-Length";
        static final String UPLOAD_HEADER_KEY_ACCEPT = "Accept";
        static final String RESPONSE_KEY_BLOB_ID = "id";
        static final String RESPONSE_KEY_ERROR = "error";

        private BlobKeys() {}
    }

    static final class PayloadDataKeys {
        static final String ANALYTICS_DEBUG_API_ENABLED = "analytics.debugApiEnabled";
        static final String STATE_DATA = "state.data";
        static final String XDM_STATE_DATA = "xdm.state.data";
        static final String METADATA = "metadata";
        static final String TYPE = "type";
        static final String DETAIL = "detail";

        private PayloadDataKeys() {}
    }

    static final class ClientInfoKeys {
        static final String VERSION = "version";
        static final String DEVICE_INFO = "deviceInfo";
        static final String APP_SETTINGS = "appSettings";

        private ClientInfoKeys() {}
    }

    static final class DeviceInfoKeys {
        static final String PLATFORM_NAME = "Canonical platform name";
        static final String DEVICE_NAME = "Device name";
        static final String DEVICE_MANUFACTURER = "Device manufacturer";
        static final String OPERATING_SYSTEM = "Operating system";
        static final String CARRIER_NAME = "Carrier name";
        static final String DEVICE_TYPE = "Device type";
        static final String MODEL = "Model";
        static final String SCREEN_SIZE = "Screen size";
        static final String LOCATION_SERVICE_ENABLED = "Location service enabled";
        static final String LOCATION_AUTHORIZATION_STATUS = "Location authorization status";
        static final String LOW_POWER_BATTERY_ENABLED = "Low power mode enabled";
        static final String BATTERY_LEVEL = "Battery level";
    }

    static final class SocketURLKeys {
        static final String SESSION_ID = "sessionId";
        static final String CLIENT_ID = "clientId";
        static final String ORG_ID = "orgId";
        static final String TOKEN = "token";
    }

    static final class SocketCloseCode {
        /**
         * Represents an the result of a successful intentional closure. Example: the user clicking
         * the disconnect button in the UI.
         */
        static final int NORMAL = 1000;

        /** Represents a class of unexpected errors. */
        static final int ABNORMAL = 1006;

        /**
         * Represents a client error during socket connection resulting due the socket connection
         * specifying an incorrect organization identifier.
         */
        static final int ORG_MISMATCH = 4900;

        /**
         * Represents a client error during socket connection Example: If clientInfoEvent is not the
         * first event to socket or, if there are any missing parameters in the socket URL.
         */
        static final int CLIENT_ERROR = 4400;

        /**
         * Represents an error resulting due to the number of connections per session exceeding the
         * default value - 200.
         */
        static final int CONNECTION_LIMIT = 4901;

        /**
         * Represents an error resulting due to the number of Assurance events sent per minute. The
         * default limit is 10k events per minute.
         */
        static final int EVENT_LIMIT = 4902;

        /**
         * Represents an error resulting due to the client attempting to connect to a session that
         * has been deleted (session may have been deleted manually, or due to exceeding the 30 day
         * limit)
         */
        static final int SESSION_DELETED = 4903;

        private SocketCloseCode() {}
    }

    static final class IntentExtraKey {
        static final String ERROR_NAME = "errorName";
        static final String ERROR_DESCRIPTION = "errorDescription";

        private IntentExtraKey() {}
    }

    // ========================================================================================
    // Enums
    // ========================================================================================

    enum AssuranceSocketError {
        GENERIC_ERROR(
                "Connection Error",
                "The connection may be failing due to a network issue or an incorrect PIN. "
                        + "Please verify internet connectivity or the PIN and try again."),
        NO_ORGID(
                "Invalid Configuration",
                "The Experience Cloud organization identifier is unavailable from the SDK. Ensure"
                    + " SDK configuration is setup correctly. See documentation for more detail."),
        ORGID_MISMATCH(
                "Unauthorized Access",
                "The Experience Cloud organization identifier does not match with that of the"
                    + " Assurance session. Ensure the right Experience Cloud organization is being"
                    + " used. See documentation for more detail."),
        CONNECTION_LIMIT(
                "Connection Limit Reached",
                "You have reached the maximum number of connected devices allowed for a session. "
                        + "Please disconnect another device and try again."),
        EVENT_LIMIT(
                "Event Limit Reached",
                "You have reached the maximum number of events that can be sent per minute."),
        CLIENT_ERROR(
                "Client Disconnected",
                "This client has been disconnected due to an unexpected error. Error Code 4400."),
        SESSION_DELETED(
                "Session Deleted",
                "The session client connected to has been deleted. Error Code 4903.");

        private final String error;
        private final String errorDescription;

        private AssuranceSocketError(final String error, final String description) {
            this.error = error;
            this.errorDescription = description;
        }

        String getErrorDescription() {
            return errorDescription;
        }

        String getError() {
            return error;
        }

        @Override
        public String toString() {
            return error + ": " + errorDescription;
        }
    }

    enum UILogColorVisibility {
        LOW(0),
        NORMAL(1),
        HIGH(2),
        CRITICAL(3);
        private final int intValue;

        UILogColorVisibility(final int val) {
            this.intValue = val;
        }

        public int getValue() {
            return intValue;
        }
    }

    enum AssuranceEnvironment {
        PROD("prod"),
        STAGE("stage"),
        QA("qa"),
        DEV("dev");

        private final String environmentString;

        /**
         * Returns {@link AssuranceEnvironment} value of the provided string.
         *
         * <p>Returns null if the provided string is not a valid {@link AssuranceEnvironment} enum
         * value
         *
         * @return {@link AssuranceEnvironment} value for provided status string
         */
        AssuranceEnvironment(final String environmentString) {
            this.environmentString = environmentString;
        }

        public String stringValue() {
            return environmentString;
        }

        /**
         * Returns {@link AssuranceEnvironment} enum value for the provided string.
         *
         * <p>Defaults to PROD if the provided string is not a valid {@link AssuranceEnvironment}
         * enum value
         *
         * @return {@link AssuranceEnvironment} value for provided status string
         */
        public static AssuranceEnvironment get(final String statusString) {
            AssuranceEnvironment enumValue = lookup.get(statusString);
            return (enumValue == null) ? AssuranceEnvironment.PROD : enumValue;
        }

        // generate look up table on load time
        private static final Map<String, AssuranceEnvironment> lookup = new HashMap<>();

        static {
            for (AssuranceEnvironment env : AssuranceEnvironment.values()) {
                lookup.put(env.stringValue(), env);
            }
        }
    }
}

File: code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssuranceConstants.kt
/*
  Copyright 2022 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
*/

package com.adobe.marketing.mobile.assurance.internal

import com.adobe.marketing.mobile.assurance.R
import com.adobe.marketing.mobile.services.ServiceProvider
import java.util.concurrent.TimeUnit

internal object AssuranceConstants {
    const val VENDOR_ASSURANCE_MOBILE = "com.adobe.griffon.mobile"

    internal object DeeplinkURLKeys {
        const val START_URL_QUERY_KEY_SESSION_ID = "adb_validation_sessionid"
        const val START_URL_QUERY_KEY_ENVIRONMENT = "env"
    }

    internal object GenericEventPayloadKey {
        const val ACP_EXTENSION_EVENT_TYPE = "ACPExtensionEventType"
        const val ACP_EXTENSION_EVENT_SOURCE = "ACPExtensionEventSource"
        const val ACP_EXTENSION_EVENT_NAME = "ACPExtensionEventName"
        const val ACP_EXTENSION_EVENT_DATA = "ACPExtensionEventData"
        const val ACP_EXTENSION_EVENT_UNIQUE_IDENTIFIER = "ACPExtensionEventUniqueIdentifier"
        const val ACP_EXTENSION_EVENT_PARENT_IDENTIFIER = "ACPExtensionEventParentIdentifier"
    }

    internal object SDKEventDataKey {
        const val START_SESSION_URL = "startSessionURL"
        const val IS_QUICK_CONNECT = "quickConnect"
        const val END_SESSION = "endSession"
        const val EXTENSIONS = "extensions"
        const val STATE_OWNER = "stateowner"
        const val FRIENDLY_NAME = "friendlyName"
    }

    internal object SDKEventName {
        const val XDM_SHARED_STATE_CHANGE = "Shared state change (XDM)"
        const val SHARED_STATE_CHANGE = "Shared state change"
    }

    internal object SDKSharedStateName {
        const val CONFIGURATION = "com.adobe.module.configuration"
        const val EVENTHUB = "com.adobe.module.eventhub"
    }

    internal object SDKConfigurationKey {
        const val ORG_ID = "experienceCloud.org"
    }

    internal object AssuranceEventType {
        const val GENERIC = "generic"
        const val LOG = "log"
        const val CONTROL = "control"
        const val CLIENT = "client"
        const val BLOB = "blob"
    }

    internal object ControlType {
        const val START_EVENT_FORWARDING = "startEventForwarding"
        const val SCREENSHOT = "screenshot"
        const val LOG_FORWARDING = "logForwarding"
        const val FAKE_EVENT = "fakeEvent"
        const val CONFIG_UPDATE = "configUpdate"
        const val NONE = "none"
        const val WILDCARD = "wildcard"
    }

    internal object AssuranceEventKeys {
        const val EVENT_ID = "eventID"
        const val VENDOR = "vendor"
        const val TYPE = "type"
        const val PAYLOAD = "payload"
        const val METADATA = "metadata"
        const val TIMESTAMP = "timestamp"
        const val EVENT_NUMBER = "eventNumber"
        const val CHUNK_DATA = "chunkData"
        const val CHUNK_ID = "chunkId"
        const val CHUNK_TOTAL = "chunkTotal"
        const val CHUNK_SEQUENCE_NUMBER = "chunkSequenceNumber"
    }

    internal object SharedStateKeys {
        const val ASSURANCE_STATE_SESSION_ID = "sessionid"
        const val ASSURANCE_STATE_CLIENT_ID = "clientid"
        const val ASSURANCE_STATE_INTEGRATION_ID = "integrationid"
    }

    internal object DataStoreKeys {
        const val DATASTORE_NAME = "com.adobe.assurance.preferences"
        const val SESSION_URL = "reconnection.url"
        const val ENVIRONMENT = "environment"
        const val CLIENT_ID = "clientid"
        const val SESSION_ID = "sessionid"
        const val TOKEN = "token"
    }

    internal object BlobKeys {
        const val UPLOAD_ENDPOINT_FORMAT = "https://blob%s.griffon.adobe.com"
        const val UPLOAD_PATH_API = "api"
        const val UPLOAD_PATH_FILEUPLOAD = "FileUpload"
        const val UPLOAD_QUERY_KEY = "validationSessionId"
        const val UPLOAD_HTTP_METHOD = "POST"
        const val UPLOAD_HEADER_KEY_CONTENT_TYPE = "Content-Type"
        const val UPLOAD_HEADER_KEY_FILE_CONTENT_TYPE = "File-Content-Type"
        const val UPLOAD_HEADER_KEY_CONTENT_LENGTH = "Content-Length"
        const val UPLOAD_HEADER_KEY_ACCEPT = "Accept"
        const val RESPONSE_KEY_BLOB_ID = "id"
        const val RESPONSE_KEY_ERROR = "error"
    }

    internal object PayloadDataKeys {
        const val ANALYTICS_DEBUG_API_ENABLED = "analytics.debugApiEnabled"
        const val STATE_DATA = "state.data"
        const val XDM_STATE_DATA = "xdm.state.data"
        const val METADATA = "metadata"
        const val TYPE = "type"
        const val DETAIL = "detail"
    }

    internal object ClientInfoKeys {
        const val VERSION = "version"
        const val DEVICE_INFO = "deviceInfo"
        const val APP_SETTINGS = "appSettings"
    }

    internal object DeviceInfoKeys {
        const val PLATFORM_NAME = "Canonical platform name"
        const val DEVICE_NAME = "Device name"
        const val DEVICE_MANUFACTURER = "Device manufacturer"
        const val OPERATING_SYSTEM = "Operating system"
        const val CARRIER_NAME = "Carrier name"
        const val DEVICE_TYPE = "Device type"
        const val MODEL = "Model"
        const val SCREEN_SIZE = "Screen size"
        const val LOCATION_SERVICE_ENABLED = "Location service enabled"
        const val LOCATION_AUTHORIZATION_STATUS = "Location authorization status"
        const val LOW_POWER_BATTERY_ENABLED = "Low power mode enabled"
        const val BATTERY_LEVEL = "Battery level"
    }

    internal object SocketURLKeys {
        const val SESSION_ID = "sessionId"
        const val CLIENT_ID = "clientId"
        const val ORG_ID = "orgId"
        const val TOKEN = "token"
    }

    internal object SocketCloseCode {
        /**
         * Represents an the result of a successful intentional closure. Example: the user clicking
         * the disconnect button in the UI.
         */
        const val NORMAL = 1000

        /** Represents a class of unexpected errors.  */
        const val ABNORMAL = 1006

        /**
         * Represents a client error during socket connection resulting due the socket connection
         * specifying an incorrect organization identifier.
         */
        const val ORG_MISMATCH = 4900

        /**
         * Represents a client error during socket connection Example: If clientInfoEvent is not the
         * first event to socket or, if there are any missing parameters in the socket URL.
         */
        const val CLIENT_ERROR = 4400

        /**
         * Represents an error resulting due to the number of connections per session exceeding the
         * default value - 200.
         */
        const val CONNECTION_LIMIT = 4901

        /**
         * Represents an error resulting due to the number of Assurance events sent per minute. The
         * default limit is 10k events per minute.
         */
        const val EVENT_LIMIT = 4902

        /**
         * Represents an error resulting due to the client attempting to connect to a session that
         * has been deleted (session may have been deleted manually, or due to exceeding the 30 day
         * limit)
         */
        const val SESSION_DELETED = 4903

        /**
         * Converts a socket close code to an `AssuranceConnectionError` if such a mapping
         * exists. Not all socket close codes are error codes and not all AssuranceConnectionErrors
         * are socket errors. So this utility is needed to bridge socket codes and
         * AssuranceConnectionError.
         *
         * @param closeCode a socket close code for which an AssuranceConnectionError is needed
         * @return an `AssuranceConnectionError`
         */
        @JvmStatic
        fun toAssuranceConnectionError(closeCode: Int): AssuranceConnectionError? {
            return when (closeCode) {
                ORG_MISMATCH -> AssuranceConnectionError.ORG_ID_MISMATCH
                CLIENT_ERROR -> AssuranceConnectionError.CLIENT_ERROR
                CONNECTION_LIMIT -> AssuranceConnectionError.CONNECTION_LIMIT
                EVENT_LIMIT -> AssuranceConnectionError.EVENT_LIMIT
                SESSION_DELETED -> AssuranceConnectionError.SESSION_DELETED
                ABNORMAL -> AssuranceConnectionError.GENERIC_ERROR
                else -> null
            }
        }
    }

    internal object QuickConnect {
        const val BASE_DEVICE_API_URL = "https://device.griffon.adobe.com/device"
        const val DEVICE_API_PATH_CREATE = "create"
        const val DEVICE_API_PATH_STATUS = "status"
        const val KEY_SESSION_ID = "sessionUuid"
        const val KEY_SESSION_TOKEN = "token"
        const val KEY_ORG_ID = "orgId"
        const val KEY_DEVICE_NAME = "deviceName"
        const val KEY_CLIENT_ID = "clientId"
        val CONNECTION_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(5).toInt()
        val READ_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(5).toInt()
        val STATUS_CHECK_DELAY_MS = TimeUnit.SECONDS.toMillis(2)
        const val MAX_RETRY_COUNT = 300
    }

    internal enum class AssuranceConnectionError(
        private val errorResId: Int,
        private val descriptionResId: Int,
        @JvmField val isRetryable: Boolean
    ) {
        GENERIC_ERROR(
            R.string.error_title_incorrect_pin_or_network,
            R.string.error_desc_incorrect_pin_or_network,
            true
        ),
        NO_ORG_ID(
            R.string.error_title_invalid_org_id,
            R.string.error_desc_invalid_org_id,
            false
        ),
        ORG_ID_MISMATCH(
            R.string.error_title_unauthorized_access,
            R.string.error_desc_unauthorized_access,
            false
        ),
        CONNECTION_LIMIT(
            R.string.error_title_connection_limit,
            R.string.error_desc_connection_limit,
            false
        ),
        EVENT_LIMIT(
            R.string.error_title_event_limit,
            R.string.error_desc_event_limit,
            false
        ),
        CLIENT_ERROR(
            R.string.error_title_unexpected_error,
            R.string.error_desc_unexpected_error,
            false
        ),
        SESSION_DELETED(
            R.string.error_title_session_deleted,
            R.string.error_desc_session_deleted,
            false
        ),
        CREATE_DEVICE_REQUEST_MALFORMED(
            R.string.error_title_invalid_registration_request,
            R.string.error_desc_invalid_registration_request,
            false
        ),
        STATUS_CHECK_REQUEST_MALFORMED(
            R.string.error_title_invalid_registration_request,
            R.string.error_desc_invalid_registration_request,
            false
        ),
        RETRY_LIMIT_REACHED(
            R.string.error_title_retry_limit_reached,
            R.string.error_desc_retry_limit_reached,
            true
        ),
        CREATE_DEVICE_REQUEST_FAILED(
            R.string.error_title_registration_error,
            R.string.error_desc_registration_error,
            true
        ),
        DEVICE_STATUS_REQUEST_FAILED(
            R.string.error_title_registration_error,
            R.string.error_desc_registration_error,
            true
        ),
        UNEXPECTED_ERROR(
            R.string.error_title_invalid_registration_response,
            R.string.error_desc_invalid_registration_response,
            true
        );

        val error: String
            get() {
                val context = ServiceProvider.getInstance().appContextService.applicationContext
                // This code path should never be crossed if the app context is null because the SDK
                // operations require a valid context. Default to empty string if context is null
                // because we don't want to crash the app.
                return context?.getString(errorResId) ?: ""
            }

        val description: String
            get() {
                val context = ServiceProvider.getInstance().appContextService.applicationContext
                // This code path should never be crossed if the app context is null because the SDK
                // operations require a valid context. Default to empty string if context is null
                // because we don't want to crash the app.
                return context?.getString(descriptionResId) ?: ""
            }
    }

    internal enum class UILogColorVisibility(val value: Int) {
        LOW(0),
        NORMAL(1),
        HIGH(2),
        CRITICAL(3)
    }

    /**
     * Returns [AssuranceEnvironment] value of the provided string.
     *
     *
     * Returns null if the provided string is not a valid [AssuranceEnvironment] enum
     * value
     *
     * @return [AssuranceEnvironment] value for provided status string
     */
    internal enum class AssuranceEnvironment(@JvmField val stringValue: String) {
        PROD("prod"),
        STAGE("stage"),
        QA("qa"),
        DEV("dev");

        /**
         * Returns [AssuranceEnvironment] enum value for the provided string.
         *
         * Defaults to PROD if the provided string is not a valid [AssuranceEnvironment]
         * enum value
         *
         * @return [AssuranceEnvironment] value for provided status string
         */
        internal companion object {
            private val lookup: Map<String, AssuranceEnvironment> = values().associateBy { it.stringValue }

            @JvmName("get")
            @JvmStatic
            internal operator fun get(stringValue: String): AssuranceEnvironment {
                val enumValue = lookup[stringValue]
                return enumValue ?: PROD
            }
        }
    }
}
